scripts/state_space_models_lib.py

```diff
@@ -0,0 +1,316 @@
+        mu0 = self.initial_state_prior.mean()
+        Sigma0 = self.initial_state_prior.covariance()
+        _, (filtered_means, filtered_covs, mu_cond_hist, Sigma_cond_hist) = jax.lax.scan(self.__kalman_step, (mu0, Sigma0), x)
+        
+        return filtered_means, filtered_covs, mu_cond_hist, Sigma_cond_hist
+    
+    def __backward_smoothing_pass(self, filtered_means: Array, filtered_covs: Array,
+                                  mu_cond_hist: Array, Sigma_cond_hist: Array) -> Tuple[Array, Array]:
+        # filtered_means, filtered_covs, mu_cond_hist, Sigma_cond_hist = self.__forward_filter(x)
+        mut_giv_T = filtered_means[-1, :]
+        Sigmat_giv_T = filtered_covs[-1, :]
+
+        elements = (filtered_means[-2::-1], filtered_covs[-2::-1, ...], mu_cond_hist[1:][::-1, ...], Sigma_cond_hist[1:][::-1, ...])
+        _, (smoothed_means, smoothed_covs) = jax.lax.scan(self.__smoother_step, (mut_giv_T, Sigmat_giv_T), elements)
+        smoothed_means = jnp.concatenate([smoothed_means[::-1, ...], mut_giv_T[None, ...]], axis=0)
+        smoothed_covs = jnp.concatenate([smoothed_covs[::-1, ...], Sigmat_giv_T[None, ...]], axis=0)
+
+        return smoothed_means, smoothed_covs
+
+    def __sample_step(self, state_t: Array, key: PRNGKey):
+        key_state, key_obs = random.split(key)
+        state_mean_next = self.transition_matrix @ state_t
+
+        state_noise = self.transition_noise.sample(seed=key_state)
+        obs_noise = self.observation_noise.sample(seed=key_obs) 
+
+        state_next = state_mean_next + state_noise
+        obs_next = self.observation_matrix @ state_next + obs_noise
+
+        return state_next, (state_next, obs_next)
+    
+    def _sample_n(self, key: PRNGKey, n: int) -> Tuple[Array, Array]:
+        key_init, key_next, key_step = random.split(key, 3)
+        key_steps = random.split(key_step, n * self.num_timesteps).reshape(n, self.num_timesteps, -1)
+        
+        state_init = self.initial_state_prior.sample(seed=key_init, sample_shape=(n,))
+        
+        vmap_sample = jax.vmap(lambda states, keys: jax.lax.scan(self.__sample_step, states, keys))
+        _, (state_samples, obs_samples) = vmap_sample(state_init, key_steps)
+
+        return state_samples, obs_samples
+
+    def sample(self, *,
+               seed: Union[IntLike, PRNGKey],
+               sample_shape: Union[IntLike, Sequence[IntLike]] = ()) -> Array:
+        """Samples an event.
+        Parameters
+        ----------
+        seed: PRNG key or integer seed.
+        sample_shape: Additional leading dimensions for sample.
+
+        Returns
+        -------
+        * Array(*sample_shape, batch_shape, event_shape)
+            A sample of shape `sample_shape` + `batch_shape` + `event_shape`.
+        """
+        rng, sample_shape = convert_seed_and_sample_shape(seed, sample_shape)
+        num_samples = functools.reduce(operator.mul, sample_shape, 1)  # product
```

Comment:
I took this function directly from the class `Distribution` in [distribution.py](https://github.com/deepmind/distrax/blob/6da616249953dbf6aaefaeed4a8c4394e64887e4/distrax/_src/distributions/distribution.py#L115) in the Distrax library. I was originally inheriting from `distrax.Distribution`, which required me to implement the `_sample_n` method, and the `sample` method called `sample_n`. Since we know inherit from `jittable.Jittable` I took it from their class